### PR TITLE
devmapper: store base device fs type

### PR DESCRIPTION
**Pull request by runcom**
## TL;DR;

After the very first init of the graph `docker info` correctly shows the
base fs type under `Backing Filesystem`. This information isn't stored
anywhere. After a restart (w/o erasing `/var/lib/docker`) `docker info`
shows an empty string under `Backing Filesystem`.
This patch records the base fs type after the first run in the metadata
or, to fix old devices that don't have this info in the metadata, just
probe the fs type of the base device at graph startup.
## Long story [(c) @calavera] 

`docker info` only shows `Backing Filesystem` the first time you start a daemon. Every later restart makes the value became empty.

First time you run a daemon:

```
[root@localhost docker]# docker info
Containers: 0
Images: 0
Server Version: 1.10.0-dev
Storage Driver: devicemapper
 Pool Name: docker-252:3-8464477-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 53.74 MB
 Data Space Total: 107.4 GB
 Data Space Available: 3.213 GB
 Metadata Space Used: 606.2 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/0.0/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/0.0/devicemapper/devicemapper/metadata
 Library Version: 1.02.113 (2015-12-05)
Execution Driver: native-0.2
Logging Driver: journald
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.0-0.rc4.git2.1.fc24.x86_64
Operating System: Fedora 24 (Rawhide)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 0
CPUs: 1
Total Memory: 973.2 MiB
Name: localhost.localdomain
ID: 4G4C:VRMN:JJSZ:2MVU:PCQD:WQI3:ZD7Y:DBN5:XMMV:UH6G:KX2J:KYPR
Experimental: true
Registries: 127.0.0.0/8 (insecure), docker.io (secure)
```

Restart the daemon (notice how Backing Filesystem is emtpy):

```
[root@localhost docker]# docker info
Containers: 0
Images: 0
Server Version: 1.10.0-dev
Storage Driver: devicemapper
 Pool Name: docker-252:3-8464477-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem:
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 53.74 MB
 Data Space Total: 107.4 GB
 Data Space Available: 3.213 GB
 Metadata Space Used: 606.2 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/0.0/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/0.0/devicemapper/devicemapper/metadata
 Library Version: 1.02.113 (2015-12-05)
Execution Driver: native-0.2
Logging Driver: journald
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.0-0.rc4.git2.1.fc24.x86_64
Operating System: Fedora 24 (Rawhide)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 0
CPUs: 1
Total Memory: 973.2 MiB
Name: localhost.localdomain
ID: 4G4C:VRMN:JJSZ:2MVU:PCQD:WQI3:ZD7Y:DBN5:XMMV:UH6G:KX2J:KYPR
Experimental: true
Registries: 127.0.0.0/8 (insecure), docker.io (secure)
```
## Reproduction

OS: Fedora Rawhide which just use devmapper as default and xfs

1) `dnf install docker` (this will install out latest fedora-1.10 branch)
2) `systemctl start docker`
3) `docker info` (this will show xfs as backing filesystem)
4) `systemctl restart docker`
5) `docker info` (this WON'T show xfs as backing filesystem)

You can actually reproduce it even after the very first daemon startup:

1) `systemctl stop docker`
2) `rm -rf /var/lib/docker`
3) `systemctl start docker`
4) `docker info` (this will show xfs as backing filesystem)
5) `systemctl restart docker`
6) `docker info` (this WON'T show xfs as backing filesystem)

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

**Comment by runcom** [comment]
ping @rhvgoyal

**Comment by calavera** [comment]
:ok_hand: LGMT

**Comment by rhvgoyal** [comment]
We probably will have to introduce another field say devices.BaseDeviceFilesystem instead of using devices.filesystem.

Right now devices.filesystem is being used to keep track of filesystem user has specified using dm.fs option. So say following happens.

- Initially dm.fs is not specified and docker creates an fs xfs.
- User shuts down docker, adds dm.fs=ext4 and restarts.
- User will get warning that dm.fs will be ignored as base device has xfs filesystem.

We don't want to break this warning part and that means devices.filesystem will have to be loaded from dm.fs and not from saved metadata. That means we will need another field to keep track of fs on base device.

**Comment by runcom** [comment]
Thanks @rhvgoyal 
As previously discussed I'll update the patch accordingly

**Comment by runcom** [comment]
@rhvgoyal updated and much cleaner now, PTAL

**Comment by runcom** [comment]
@rhvgoyal updated again with your suggestions

**Comment by rhvgoyal** [comment]
@runcom thanks for the changes.

LGTM

**Comment by runcom** [comment]
@calavera PTAL

**Comment by vdemeester** [comment]
LGTM 🐯

**Comment by vdemeester** [comment]
Merging, devmapper and windows are not really related :wink:
